server/src/main/java/org/elasticsearch/search/lookup/ReinitializingSourceProvider.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.search.lookup;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * This is a workaround for when compute engine executes concurrently with data partitioning by docid.
+ */
+public class ReinitializingSourceProvider implements SourceProvider {
+
+    private final Supplier<SourceProvider> sourceProviderFactory;
+    private final Map<Long, SourceProvider> map = ConcurrentCollections.newConcurrentMap();
+
+    public ReinitializingSourceProvider(Supplier<SourceProvider> sourceProviderFactory) {
+        this.sourceProviderFactory = sourceProviderFactory;
+    }
+
+    @Override
+    public Source getSource(LeafReaderContext ctx, int doc) throws IOException {
+        var currentThread = Thread.currentThread();
+        var sourceProvider = map.computeIfAbsent(currentThread.threadId(), (key) -> sourceProviderFactory.get());
```

Comment:
This could result in a memory leak since the map is never cleaned. Another option would be to remove items when a thread ends, but that would be too complex. For now, I think it's okay to keep the most recently created source provider.

```
    private PerThreadSourceProvider perThreadProvider;
    private final Supplier<SourceProvider> sourceProviderFactory;

    public ReinitializingSourceProvider(Supplier<SourceProvider> sourceProviderFactory) {
        this.sourceProviderFactory = sourceProviderFactory;
    }

    @Override
    public Source getSource(LeafReaderContext ctx, int doc) throws IOException {
        var currentThread = Thread.currentThread();
        PerThreadSourceProvider provider = perThreadProvider;
        if (provider == null || provider.creatingThread != currentThread) {
            provider = new PerThreadSourceProvider(sourceProviderFactory.get(), currentThread);
            this.perThreadProvider = provider;
        }
        return perThreadProvider.source.getSource(ctx, doc);
    }

    private record PerThreadSourceProvider(SourceProvider source, Thread creatingThread) {

    }
```
WDYT?

Reply:
I discussed this with Martijn offline — the map is cleaned up once the query completes.